fix(sync): SAV-947: Take advisory lock -before- updating sync tick

edmofro opened the pull request:
### Changes

See comment inline

### Deploys

- [ ] **Deploy to Tamanu Internal** <!-- #deploy -->

### Tests

- [ ] **Run E2E Tests** <!-- #e2e -->

### Remember to...

- ...write or update tests
- ...add UI screenshots and **testing notes** to the Linear issue
- ...add any **manual upgrade steps** to the Linear issue
- ...update the [config reference](https://beyond-essential.slab.com/posts/reference-config-file-0c70ukly), [settings reference](https://beyond-essential.slab.com/posts/reference-settings-0blw1x2q), or any [relevant runbook(s)](https://beyond-essential.slab.com/topics/runbooks-bs04ml6c)
- ...call out additions or changes to **config files** for the deployment team to take note of

<!-- Thank you! -->

edmofro commented on packages/database/src/migrations/1744340076240-fixRaceConditionInSettingUpdateSyncTick.ts:
```diff
@@ -0,0 +1,67 @@
+        IF NEW.updated_at_sync_tick = -1 THEN
+          NEW.updated_at_sync_tick := -999;
+        ELSE
+          -- take an advisory lock on the current sync tick (if one doesn't already exist), to
+          -- record that an active transaction is using this sync tick
+          -- see waitForPendingEditsUsingSyncTick for more details
+          PERFORM pg_try_advisory_xact_lock_shared(NEW.updated_at_sync_tick);
+
+          -- now that the advisory lock is acquired, we can safely select the current sync tick and
+          -- know that no sync session will skip over this change until the transaction is complete
+          SELECT value FROM local_system_facts WHERE key = 'currentSyncTick' INTO NEW.updated_at_sync_tick;
```
The only difference is that these two are flipped.
Previously the advisory look was taken after the sync tick was selected, meaning there was a potential race condition where:
1. A record was changed with sync tick x
2. A sync snapshot including this sync tick started 
3. The pg advisory lock was taken
Steps 1 and 3 are the two lines of SQL visible above (in their old order). If step 2 managed to slip in between them, it could think that the snapshot had all records for sync tick x, but actually there was still an in-flight transaction with another record, so that record would be skipped by that sync session. 

In reality, this would (I think) never ever happen, because within step 2. there would be a few lines of javascript, with awaits, and an SQL query, which would be unlikely to fit between 1. and 3.

This is where we do the wait
```
      const { tick: currentTick } = await this.tickTockGlobalClock();

      await this.waitForPendingEdits(currentTick);
```
So the record save would need to happen before the tick tock first line (if it happened afterwards, it would use the "tock" of the tick tock, which would not be considered part of the range of this snapshot), but the advisory lock would need to be taken after the tick tock. 

FWIW, the tick tock code looks like this:
```
  static async incrementValue(key: FactName, amount: number = 1) {
    const [rowsAffected] = await this.sequelize.query(
      `
        UPDATE
          local_system_facts
        SET
          value = value::integer + :amount,
          updated_at = NOW()
        WHERE
          key = :key
        RETURNING
          value;
      `,
      { replacements: { key, amount } },
    );
    if (rowsAffected.length === 0) {
      throw new Error(`The local system fact table does not include the fact ${key}`);
    }
    const fact = rowsAffected[0] as LocalSystemFact;
    return fact.value;
  }
```
So it's a bit of work.

edmofro commented:
This PR is closed, but can be reopened and tested as part of the sync maintenance sprint

dannash100 reviewed (approved):
Thanks for re-grabbing this!

dannash100 reviewed (approved):
Thanks !

edmofro committed:
fix(sync): SAV-947: Fix race condition in sync tick updates

- Take advisory lock before updating sync tick to prevent race conditions
- Fix race condition in set_updated_at_sync_tick trigger
- Correct down migration order in sync tick race condition fix
- Clean up cursor-related code additions

edmofro commented:
Rebased to the epic and will now merge
